File: Python/format_with_timezone.py
import pandas as pd
from datetime import datetime
import pytz

def format_with_timezone(time, format_str="%Y-%m-%d %H:%M:%S %Z", to_tz=None):
    """
    Format a datetime object or a Series of datetime objects with time zone information.
    
    This function takes a datetime object or a pandas Series of datetime objects 
    and optionally converts them to a specified time zone before formatting them 
    according to the provided format string.
    
    Parameters:
        time (datetime or pd.Series): A datetime object or a pandas Series of datetime objects.
        format_str (str): A string specifying the desired format. Default is "%Y-%m-%d %H:%M:%S %Z".
        to_tz (str): A string specifying the target time zone. If provided,
                     the datetime object(s) are converted to this time zone before formatting.
                     Default is None (no conversion).
                     
    Returns:
        str or pd.Series: A string or a Series representing the formatted datetime object(s) 
                          with time zone information.
    
    Raises:
        ValueError: If any 'time' object is naive or if 'to_tz' is not a valid time zone.
    """
    
    def format_single_datetime(dt):
        # Check if the time object is naive (i.e., without timezone info)
        if dt.tzinfo is None:
            raise ValueError("The time object is naive. Please provide a datetime object with time zone information.")
        
        # If a target time zone is specified, validate and convert the time object to that time zone
        if to_tz:
            try:
                target_tz = pytz.timezone(to_tz)
            except pytz.UnknownTimeZoneError:
                raise ValueError(f"The to_tz '{to_tz}' is not a valid time zone.")
            
            # Convert the time object to the target time zone
            dt = dt.astimezone(target_tz)
        
        # Format the datetime object using the provided format string
        return dt.strftime(format_str)

    if isinstance(time, pd.Series):
        return time.map(format_single_datetime)
    else:
        return format_single_datetime(time)

# Example usage with DataFrame
df = pd.DataFrame({
    'time': [
        "2023-09-23 12:34:56",
        "2023-09-24 13:45:57"
    ]
})

df['time'] = pd.to_datetime(df['time']).dt.tz_localize('UTC')
df['formatted_time'] = format_with_timezone(df['time'], to_tz="America/New_York")

print(df)

File: R/format_with_timezone.R
library(lubridate)

format_with_timezone <- function(time, format_str = "%Y-%m-%d %H:%M:%S %Z", to_tz = NULL) {
  #' Format a datetime object or a column of datetime objects with time zone information.
  #'
  #' This function takes a datetime object or a column of datetime objects and
  #' optionally converts them to a specified time zone before formatting them
  #' according to the provided format string.
  #'
  #' @param time A POSIXct object or a vector of POSIXct objects.
  #' @param format_str A character string specifying the desired format.
  #'   Default is "%Y-%m-%d %H:%M:%S %Z".
  #' @param to_tz A character string specifying the target time zone. If provided,
  #'   the datetime object is converted to this time zone before formatting.
  #'   Default is NULL (no conversion).
  #'
  #' @return A character string or a character vector representing the formatted datetime object(s)
  #'   with time zone information.
  
  # Apply the formatting function to each element of the time object if it is a vector
  formatted_time <- vapply(time, function(t) {
    # Check if the time object is NA
    if (is.na(t)) {
      stop("The time object is NA. Please provide a valid POSIXct or POSIXlt object.")
    }
    
    # Check if the time object is of class POSIXct or POSIXlt
    if (!inherits(t, "POSIXt")) {
      stop("The time object must be of class POSIXct or POSIXlt.")
    }
    
    # Check if format_str is a character string
    if (!is.character(format_str) || length(format_str) != 1) {
      stop("The format_str must be a single character string.")
    }
    
    # If a target time zone is specified, validate and convert the time object to that time zone
    if (!is.null(to_tz)) {
      if (!is.character(to_tz) || length(to_tz) != 1) {
        stop("The to_tz must be a single character string representing a valid time zone.")
      }
      if (!to_tz %in% OlsonNames()) {
        stop("The to_tz is not a valid time zone.")
      }
      t <- with_tz(t, tzone = to_tz)
    }
    
    # Check if time zone information is available
    if (is.na(attr(t, "tzone"))) {
      warning("Time zone information is not available. The time object may be printed without time zone information.")
    }
    
    # Format the datetime object using the provided format string
    format(t, format_str)
  }, character(1))
  
  return(formatted_time)
}

# Example usage with a vector of POSIXct objects
time_vector <- as.POSIXct(c("2023-09-23 12:34:56", "2023-09-24 13:45:57"), tz = "UTC")
format_str <- "%Y-%m-%d %H:%M:%S %Z"
formatted_time_vector <- format_with_timezone(time_vector, format_str, to_tz = "America/New_York")
print(formatted_time_vector)